https://github.com/sihsieh65-tech/cs82a-portfolio/upload/main/Module3

Step 1: Load and size up the file

Loading the data into the analysis

In [1]:
import pandas as pd
df = pd.read_csv('messy_sales.csv')
df

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
...,...,...,...,...,...,...
295,1211,2026-05-15,charger,11.31,4,2116
296,1067,2026-04-27,backpack,37.93,5,90210
297,1025,2026-04-27,keyboard,47.30,-2,2116
298,1196,04/25/2026,pen set,8.79,3,90405


Checking the shape (.shape), first 10 rows (.head(10)), data types (.ifno()) within the data

In [2]:
df.shape

(300, 6)

In [3]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   order_id  300 non-null    int64  
 1   date      300 non-null    str    
 2   product   300 non-null    str    
 3   price     288 non-null    float64
 4   qty       300 non-null    int64  
 5   zip       300 non-null    int64  
dtypes: float64(1), int64(3), str(2)
memory usage: 14.2 KB


Step 2: Write your first observations (markdown cell)

Problems I can see in the data include:

1) The dates are acutally strings and not in a date data format
2) The dates are are not consistently formatted
3) There are instances of missing data in the price column (index lines 3 and 5)
4) There is an impossible sale quantity of -2 in index row 297



Step 3: Count the missing values

Counting missing data using .isna().sum(). We identify 12 cases in the price column only.

In [5]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

Checking the median price in the column. We will use median and not mean as the mean can skew the data towards any extremes within the data set. Checking the min and max for the price data using .describe() we note a large range between $2.93 and 479.29. Therefore the median seems more appropriate. 

In [6]:
df.describe()

,order_id,price,qty,zip
count,300.000000,288.000000,300.000000,300.000000
mean,1144.816667,56.068438,2.913333,45385.276667
std,84.515319,72.307182,1.560107,39345.608641
min,1000.000000,2.930000,-5.000000,2116.000000
25%,1071.750000,10.945000,2.000000,2134.000000
50%,1144.500000,37.530000,3.000000,30303.000000
75%,1217.250000,61.417500,4.000000,90210.000000
max,1291.000000,479.290000,5.000000,98101.000000


Step 4: Impute the missing prices

Filling in missing data with the median price.

In [7]:
df["price"].median()


np.float64(37.53)

In [8]:
df["price"] = df["price"].fillna(df["price"].median())


Checking that the prices were filled. Will likely skip this step (checking in the column only) in favor of re-examining the table.

In [9]:
df["price"]

0      11.36
1      14.79
2       5.50
3      37.53
4      45.59
       ...  
295    11.31
296    37.93
297    47.30
298     8.79
299     4.63
Name: price, Length: 300, dtype: float64

Re-examining the data in the full table.

In [10]:
df

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,37.53,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
...,...,...,...,...,...,...
295,1211,2026-05-15,charger,11.31,4,2116
296,1067,2026-04-27,backpack,37.93,5,90210
297,1025,2026-04-27,keyboard,47.30,-2,2116
298,1196,04/25/2026,pen set,8.79,3,90405


In [11]:
df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

Step 5: Remove duplicate rows

Now checking for duplicate rows with .duplicated().sum() then dropping duplicates from the table with df.drop_duplicates()

In [12]:
df.duplicated().sum()

np.int64(8)

In [13]:
df = df.drop_duplicates()

Checking that 8 rows were dropped using .shape

In [14]:
df.shape

(292, 6)

Step 6: Repair the zip codes

Correcting the zip codes to re-add leading zeroes. Set the data type to string and setting to give 5 digits, then rechecking the data by displaying it with df. 

In [15]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

In [16]:
df

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,37.53,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
...,...,...,...,...,...,...
294,1103,2026-04-21,pen set,16.47,3,90405
296,1067,2026-04-27,backpack,37.93,5,90210
297,1025,2026-04-27,keyboard,47.30,-2,02116
298,1196,04/25/2026,pen set,8.79,3,90405


Rechecking the data type with .info() to ensure that the zip is now in a str format

In [17]:
df.info()

<class 'pandas.DataFrame'>
Index: 292 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   order_id  292 non-null    int64  
 1   date      292 non-null    str    
 2   product   292 non-null    str    
 3   price     292 non-null    float64
 4   qty       292 non-null    int64  
 5   zip       292 non-null    str    
dtypes: float64(1), int64(2), str(3)
memory usage: 16.0 KB


Step 7: Standardize the dates

Changing the dates from strings to date data types and noting a "mixed" format. Then I rechecked the table by displaying with df and using .info() to make sure the datatype is correct. 

In [18]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

In [19]:
df

,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405
...,...,...,...,...,...,...
294,1103,2026-04-21,pen set,16.47,3,90405
296,1067,2026-04-27,backpack,37.93,5,90210
297,1025,2026-04-27,keyboard,47.30,-2,02116
298,1196,2026-04-25,pen set,8.79,3,90405


In [20]:
df.info()

<class 'pandas.DataFrame'>
Index: 292 entries, 0 to 299
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   order_id  292 non-null    int64         
 1   date      292 non-null    datetime64[us]
 2   product   292 non-null    str           
 3   price     292 non-null    float64       
 4   qty       292 non-null    int64         
 5   zip       292 non-null    str           
dtypes: datetime64[us](1), float64(1), int64(2), str(2)
memory usage: 16.0 KB


Step 8: Investigate the negative quantities

Identifying the incorrect negative quantities by filtering for df["qty"]<0

In [21]:
df[df["qty"]<0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


Negative values are non-sensical in this context. We cannot sell negative units. Therefore I am interpreting this to be a typo and will simply convert the numbers to positive values. 

In [22]:
df["qty"] = df["qty"].abs()

Convert negative values to positive ones using the absolute value.

In [23]:
df[df["qty"]<0]

,order_id,date,product,price,qty,zip


Verified that no negative values remain.

In [24]:
df

,order_id,date,product,price,qty,zip
0,1254,2026-04-06,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,2026-05-06,mug,5.50,1,10001
3,1066,2026-05-30,notebook,37.53,1,98101
4,1114,2026-04-06,webcam,45.59,4,90405
...,...,...,...,...,...,...
294,1103,2026-04-21,pen set,16.47,3,90405
296,1067,2026-04-27,backpack,37.93,5,90210
297,1025,2026-04-27,keyboard,47.30,2,02116
298,1196,2026-04-25,pen set,8.79,3,90405


Step 9: Save the cleaned file

Data file saved as sales_clean.csv

In [26]:
df.to_csv('sales_clean.csv', index=False)

Step 10: Write the cleaning log (markdown cell)

Cleaning Log
1. Imputed missing prices using the median (37.53) as the variance the in the data appeared large. 12 of 300 rows in the price column were corrected in this way.
2. 8 duplicate rows were removed leaving 292 rows. 
3. Zip codes were converted from integers to strings and set to contain 5 values to readd leading zeroes to the codes
4. Dates were standardized in format and changed from a string data type to a date data type.
5. 3 negative values in the quantity column were converted to positive values using the absolute value function. Negative values in this column were deemed impossible and therefore were corrected. 
   